core/history.go

```diff
@@ -0,0 +1,135 @@
+package core
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/db"
+)
+
+type History struct {
+	txn db.Transaction
+}
+
+func NewHistory(txn db.Transaction) *History {
+	return &History{txn: txn}
+}
+
+func logDBKey(key []byte, height uint64) []byte {
+	return binary.BigEndian.AppendUint64(key, height)
+}
+
+func (h *History) logOldValue(key, value []byte, height uint64) error {
+	return h.txn.Set(logDBKey(key, height), value)
+}
+
+func (h *History) deleteLog(key []byte, height uint64) error {
+	return h.txn.Delete(logDBKey(key, height))
+}
+
+func (h *History) valueAt(key []byte, height uint64) ([]byte, error) {
+	it, err := h.txn.NewIterator()
+	if err != nil {
+		return nil, err
+	}
+
+	for it.Seek(logDBKey(key, height)); it.Valid(); it.Next() {
+		seekedKey := it.Key()
+		// seekedKey size should be `len(key) + sizeof(uint64)` and seekedKey should match key prefix
+		if len(seekedKey) != len(key)+8 || !bytes.HasPrefix(seekedKey, key) {
+			break
+		}
+
+		seekedHeight := binary.BigEndian.Uint64(seekedKey[len(key):])
+		if seekedHeight < height {
+			// last change happened before the height we are looking for
+			// check head state
+			break
+		} else if seekedHeight == height {
+			// a log exists for the height we are looking for, so the old value in this log entry is not useful.
+			// advance the iterator and see we can use the next entry. If not, ErrCheckHeadState will be returned
+			continue
+		}
+
+		val, itErr := it.Value()
+		if err = db.CloseAndWrapOnError(it.Close, itErr); err != nil {
+			return nil, err
+		}
+		// seekedHeight > height
+		return val, nil
+	}
+
+	return nil, db.CloseAndWrapOnError(it.Close, errors.New("check head state"))
```

Comment:
What do you think about creating an error type for `errors.New("check head state")`? It seems to be referenced a couple of times even outside `history`

Reply:
All references are in tests if I am not mistaken. So until we match the error somewhere in production code, it should stay as a plain string error. 

I am working on a separate PR right now, that matches the "check head state" error in actual code. So we will have to make it a global error var in that PR.